daft/runners/dynamic_runner.py

```diff
@@ -0,0 +1,187 @@
+    ScheduleJoin,
+    ScheduleMaterialize,
+    SchedulePartitionRead,
+    SchedulePipelineInstruction,
+    ScheduleSort,
+)
+from daft.internal.rule_runner import FixedPointPolicy, Once, RuleBatch, RuleRunner
+from daft.logical import logical_plan
+from daft.logical.logical_plan import LogicalPlan, PartitionScheme
+from daft.logical.optimizer import (
+    DropProjections,
+    DropRepartition,
+    FoldProjections,
+    PruneColumns,
+    PushDownClausesIntoScan,
+    PushDownLimit,
+    PushDownPredicates,
+)
+from daft.runners.partitioning import PartitionCacheEntry
+from daft.runners.runner import Runner
+
+
+class DynamicRunner(Runner):
+    def __init__(self) -> None:
+        super().__init__()
+        # From PyRunner
+        self._optimizer = RuleRunner(
+            [
+                RuleBatch(
+                    "SinglePassPushDowns",
+                    Once,
+                    [
+                        DropRepartition(),
+                        PushDownPredicates(),
+                        PruneColumns(),
+                        FoldProjections(),
+                        PushDownClausesIntoScan(),
+                    ],
+                ),
+                RuleBatch(
+                    "PushDownLimitsAndRepartitions",
+                    FixedPointPolicy(3),
+                    [PushDownLimit(), DropRepartition(), DropProjections()],
+                ),
+            ]
+        )
+
+    def optimize(self, plan: LogicalPlan) -> LogicalPlan:
+        # From PyRunner
+        return self._optimizer.optimize(plan)
+
+    def run(self, plan: LogicalPlan) -> PartitionCacheEntry:
+        plan = self.optimize(plan)
+
+        schedule = schedule_logical_node(plan)
+        schedule = ScheduleMaterialize(schedule)
+
+        for next_construction in schedule:
+            if next_construction is None:
+                import IPython
+
+                IPython.embed()
+            assert next_construction is not None
+            self._build_partitions(next_construction)
+
+        final_result = schedule.result_partition_set()
+        pset_entry = self.put_partition_set_into_cache(final_result)
+        return pset_entry
+
+    def _build_partitions(self, partspec: Construction) -> None:
+        partitions = partspec.inputs
+
+        for instruction in partspec.instruction_stack:
+            partitions = instruction(partitions)
+
+        partspec.report_completed(partitions)
+
+
+def schedule_logical_node(node: LogicalPlan) -> DynamicSchedule:
+    """Translates a LogicalPlan into an appropriate DynamicSchedule."""
+
+    # -- Leaf nodes. --
+    if isinstance(node, logical_plan.InMemoryScan):
+        pset = get_context().runner().get_partition_set_from_cache(node._cache_entry.key).value
+        assert pset is not None
+        partitions = pset.values()
+        return SchedulePartitionRead(partitions)
+
+    elif isinstance(node, logical_plan.Scan):
+        return ScheduleFileRead(node)
+
```

Comment:
let's have an else case that throws a nonimplementederror, for when we implement a new logical node but it isn't registered yet.